File: speedlimit_lookup.c
/*
 * Speed Limit Lookup - Optimized C Implementation for IoT Devices
 *
 * Compile: gcc -o speedlimit speedlimit_lookup.c -lsqlite3 -lm
 * Usage: ./speedlimit za_speedlimits.db -33.9249 18.4241
 */

#include <stdio.h>
#include <stdlib.h>
#include <sqlite3.h>
#include <math.h>

typedef struct {
    double min_lat;
    double max_lat;
    double min_lon;
    double max_lon;
    int grid_size;
} DatabaseBounds;

typedef struct {
    sqlite3 *db;
    DatabaseBounds bounds;
    sqlite3_stmt *grid_stmt;
    sqlite3_stmt *bbox_stmt;
} SpeedLimitContext;

/*
 * Initialize the speed limit lookup system
 * Call once at startup
 */
int speedlimit_init(SpeedLimitContext *ctx, const char *db_path) {
    int rc;

    // Open database in read-only mode
    rc = sqlite3_open_v2(db_path, &ctx->db,
                         SQLITE_OPEN_READONLY, NULL);
    if (rc != SQLITE_OK) {
        fprintf(stderr, "Cannot open database: %s\n", sqlite3_errmsg(ctx->db));
        return rc;
    }

    // Load metadata bounds (cached for grid calculations)
    sqlite3_stmt *meta_stmt;
    const char *meta_sql =
        "SELECT "
        "  (SELECT CAST(value AS REAL) FROM metadata WHERE key = 'min_latitude'),"
        "  (SELECT CAST(value AS REAL) FROM metadata WHERE key = 'max_latitude'),"
        "  (SELECT CAST(value AS REAL) FROM metadata WHERE key = 'min_longitude'),"
        "  (SELECT CAST(value AS REAL) FROM metadata WHERE key = 'max_longitude'),"
        "  (SELECT CAST(value AS INTEGER) FROM metadata WHERE key = 'grid_size')";

    rc = sqlite3_prepare_v2(ctx->db, meta_sql, -1, &meta_stmt, NULL);
    if (rc != SQLITE_OK) return rc;

    if (sqlite3_step(meta_stmt) == SQLITE_ROW) {
        ctx->bounds.min_lat = sqlite3_column_double(meta_stmt, 0);
        ctx->bounds.max_lat = sqlite3_column_double(meta_stmt, 1);
        ctx->bounds.min_lon = sqlite3_column_double(meta_stmt, 2);
        ctx->bounds.max_lon = sqlite3_column_double(meta_stmt, 3);
        ctx->bounds.grid_size = sqlite3_column_int(meta_stmt, 4);
    }
    sqlite3_finalize(meta_stmt);

    // Prepare grid-based query (FASTEST - use this for real-time lookups)
    const char *grid_sql =
        "SELECT rs.speed_limit_kmh "
        "FROM spatial_grid sg "
        "JOIN road_segments rs ON sg.road_segment_id = rs.id "
        "WHERE sg.grid_x BETWEEN ?1 AND ?2 "
        "  AND sg.grid_y BETWEEN ?3 AND ?4 "
        "  AND rs.min_lat <= ?5 AND rs.max_lat >= ?5 "
        "  AND rs.min_lon <= ?6 AND rs.max_lon >= ?6 "
        "ORDER BY "
        "    (rs.center_lat - ?5) * (rs.center_lat - ?5) + "
        "    (rs.center_lon - ?6) * (rs.center_lon - ?6) "
        "LIMIT 1";

    rc = sqlite3_prepare_v2(ctx->db, grid_sql, -1, &ctx->grid_stmt, NULL);
    if (rc != SQLITE_OK) return rc;

    // Prepare bounding box query (FALLBACK - simpler but slower)
    const char *bbox_sql =
        "SELECT speed_limit_kmh "
        "FROM road_segments "
        "WHERE center_lat BETWEEN ?1 - 0.01 AND ?1 + 0.01 "
        "  AND center_lon BETWEEN ?2 - 0.01 AND ?2 + 0.01 "
        "ORDER BY "
        "    (center_lat - ?1) * (center_lat - ?1) + "
        "    (center_lon - ?2) * (center_lon - ?2) "
        "LIMIT 1";

    rc = sqlite3_prepare_v2(ctx->db, bbox_sql, -1, &ctx->bbox_stmt, NULL);
    if (rc != SQLITE_OK) return rc;

    return SQLITE_OK;
}

/*
 * Calculate grid coordinates from GPS position
 */
void calculate_grid_coords(SpeedLimitContext *ctx, double lat, double lon,
                          int *grid_x, int *grid_y) {
    double norm_x = (lon - ctx->bounds.min_lon) /
                    (ctx->bounds.max_lon - ctx->bounds.min_lon);
    double norm_y = (lat - ctx->bounds.min_lat) /
                    (ctx->bounds.max_lat - ctx->bounds.min_lat);

    *grid_x = (int)(norm_x * ctx->bounds.grid_size);
    *grid_y = (int)(norm_y * ctx->bounds.grid_size);

    // Clamp to valid range
    if (*grid_x < 0) *grid_x = 0;
    if (*grid_x >= ctx->bounds.grid_size) *grid_x = ctx->bounds.grid_size - 1;
    if (*grid_y < 0) *grid_y = 0;
    if (*grid_y >= ctx->bounds.grid_size) *grid_y = ctx->bounds.grid_size - 1;
}

/*
 * Lookup speed limit using grid-based query (FASTEST)
 * Returns: speed limit in km/h, or -1 if not found
 */
int speedlimit_lookup_grid(SpeedLimitContext *ctx, double lat, double lon) {
    int grid_x, grid_y;
    calculate_grid_coords(ctx, lat, lon, &grid_x, &grid_y);

    // Reset prepared statement
    sqlite3_reset(ctx->grid_stmt);
    sqlite3_clear_bindings(ctx->grid_stmt);

    // Bind parameters: grid_x-1, grid_x+1, grid_y-1, grid_y+1, lat, lon
    sqlite3_bind_int(ctx->grid_stmt, 1, grid_x - 1);
    sqlite3_bind_int(ctx->grid_stmt, 2, grid_x + 1);
    sqlite3_bind_int(ctx->grid_stmt, 3, grid_y - 1);
    sqlite3_bind_int(ctx->grid_stmt, 4, grid_y + 1);
    sqlite3_bind_double(ctx->grid_stmt, 5, lat);
    sqlite3_bind_double(ctx->grid_stmt, 6, lon);

    int speed_limit = -1;
    if (sqlite3_step(ctx->grid_stmt) == SQLITE_ROW) {
        speed_limit = sqlite3_column_int(ctx->grid_stmt, 0);
    }

    return speed_limit;
}

/*
 * Lookup speed limit using bounding box query (FALLBACK)
 * Returns: speed limit in km/h, or -1 if not found
 */
int speedlimit_lookup_bbox(SpeedLimitContext *ctx, double lat, double lon) {
    sqlite3_reset(ctx->bbox_stmt);
    sqlite3_clear_bindings(ctx->bbox_stmt);

    sqlite3_bind_double(ctx->bbox_stmt, 1, lat);
    sqlite3_bind_double(ctx->bbox_stmt, 2, lon);

    int speed_limit = -1;
    if (sqlite3_step(ctx->bbox_stmt) == SQLITE_ROW) {
        speed_limit = sqlite3_column_int(ctx->bbox_stmt, 0);
    }

    return speed_limit;
}

/*
 * Lookup speed limit (tries grid first, falls back to bbox)
 * Returns: speed limit in km/h, or -1 if not found
 */
int speedlimit_lookup(SpeedLimitContext *ctx, double lat, double lon) {
    int speed_limit = speedlimit_lookup_grid(ctx, lat, lon);

    // Fallback to bounding box if grid lookup fails
    if (speed_limit == -1) {
        speed_limit = speedlimit_lookup_bbox(ctx, lat, lon);
    }

    return speed_limit;
}

/*
 * Cleanup - call at shutdown
 */
void speedlimit_cleanup(SpeedLimitContext *ctx) {
    if (ctx->grid_stmt) sqlite3_finalize(ctx->grid_stmt);
    if (ctx->bbox_stmt) sqlite3_finalize(ctx->bbox_stmt);
    if (ctx->db) sqlite3_close(ctx->db);
}

/*
 * Example usage
 */
int main(int argc, char *argv[]) {
    if (argc != 4) {
        printf("Usage: %s <database.db> <latitude> <longitude>\n", argv[0]);
        printf("Example: %s Database/za_speedlimits.db -33.9249 18.4241\n", argv[0]);
        return 1;
    }

    const char *db_path = argv[1];
    double lat = atof(argv[2]);
    double lon = atof(argv[3]);

    SpeedLimitContext ctx = {0};

    // Initialize (do this once at startup)
    if (speedlimit_init(&ctx, db_path) != SQLITE_OK) {
        fprintf(stderr, "Failed to initialize\n");
        return 1;
    }

    printf("Looking up speed limit for: %.6f, %.6f\n", lat, lon);

    // Lookup speed limit
    int speed_limit = speedlimit_lookup(&ctx, lat, lon);

    if (speed_limit != -1) {
        printf("Speed limit: %d km/h\n", speed_limit);
    } else {
        printf("No road found at this location\n");
    }

    // Cleanup
    speedlimit_cleanup(&ctx);

    return 0;
}

File: SpeedLimits.Core/SpeedLimitLookup.cs
/*
 * Speed Limit Lookup - Optimized C# Implementation for .NET IoT
 *
 * Usage:
 *   var lookup = new SpeedLimitLookup("za_speedlimits.db");
 *   int speedLimit = lookup.GetSpeedLimit(-33.9249, 18.4241);
 *   Console.WriteLine($"Speed limit: {speedLimit} km/h");
 */

using Microsoft.Data.Sqlite;
using System;

namespace SpeedLimits.Core
{
    /// <summary>
    /// Fast speed limit lookup optimized for IoT devices
    /// </summary>
    public class SpeedLimitLookup : IDisposable
    {
        private readonly SqliteConnection _connection;
        private readonly SqliteCommand _gridCommand;
        private readonly SqliteCommand _bboxCommand;
        private readonly DatabaseBounds _bounds;
        private readonly bool _hasPlacesTable;
        private readonly SqliteCommand? _streetCmd;
        private readonly SqliteCommand? _suburbCmd;
        private readonly SqliteCommand? _cityCmd;

        private struct DatabaseBounds
        {
            public double MinLatitude;
            public double MaxLatitude;
            public double MinLongitude;
            public double MaxLongitude;
            public int GridSize;
        }

        /// <summary>
        /// Initialize speed limit lookup system (call once at startup)
        /// </summary>
        public SpeedLimitLookup(string databasePath)
        {
            // Open database in read-only mode
            _connection = new SqliteConnection($"Data Source={databasePath};Mode=ReadOnly");
            _connection.Open();

            // Load and cache metadata bounds
            _bounds = LoadBounds();

            // Prepare grid-based query (FASTEST)
            _gridCommand = _connection.CreateCommand();
            _gridCommand.CommandText = @"
                SELECT rs.speed_limit_kmh
                FROM spatial_grid sg
                JOIN road_segments rs ON sg.road_segment_id = rs.id
                WHERE sg.grid_x BETWEEN @gridXMin AND @gridXMax
                  AND sg.grid_y BETWEEN @gridYMin AND @gridYMax
                  AND rs.min_lat <= @lat AND rs.max_lat >= @lat
                  AND rs.min_lon <= @lon AND rs.max_lon >= @lon
                ORDER BY
                    (rs.center_lat - @lat) * (rs.center_lat - @lat) +
                    (rs.center_lon - @lon) * (rs.center_lon - @lon)
                LIMIT 1";

            _gridCommand.Parameters.Add("@gridXMin", SqliteType.Integer);
            _gridCommand.Parameters.Add("@gridXMax", SqliteType.Integer);
            _gridCommand.Parameters.Add("@gridYMin", SqliteType.Integer);
            _gridCommand.Parameters.Add("@gridYMax", SqliteType.Integer);
            _gridCommand.Parameters.Add("@lat", SqliteType.Real);
            _gridCommand.Parameters.Add("@lon", SqliteType.Real);

            // Prepare bounding box query (FALLBACK)
            _bboxCommand = _connection.CreateCommand();
            _bboxCommand.CommandText = @"
                SELECT speed_limit_kmh
                FROM road_segments
                WHERE center_lat BETWEEN @lat - 0.01 AND @lat + 0.01
                  AND center_lon BETWEEN @lon - 0.01 AND @lon + 0.01
                ORDER BY
                    (center_lat - @lat) * (center_lat - @lat) +
                    (center_lon - @lon) * (center_lon - @lon)
                LIMIT 1";

            _bboxCommand.Parameters.Add("@lat", SqliteType.Real);
            _bboxCommand.Parameters.Add("@lon", SqliteType.Real);

            // Check for places table (backward compatible)
            _hasPlacesTable = TableExists("places");

            if (_hasPlacesTable)
            {
                _streetCmd = _connection.CreateCommand();
                _streetCmd.CommandText = @"
                    SELECT name, highway_type
                    FROM road_segments
                    WHERE name IS NOT NULL
                      AND center_lat BETWEEN @lat - 0.005 AND @lat + 0.005
                      AND center_lon BETWEEN @lon - 0.005 AND @lon + 0.005
                    ORDER BY
                        (center_lat - @lat) * (center_lat - @lat) +
                        (center_lon - @lon) * (center_lon - @lon)
                    LIMIT 1";
                _streetCmd.Parameters.Add("@lat", SqliteType.Real);
                _streetCmd.Parameters.Add("@lon", SqliteType.Real);

                _suburbCmd = _connection.CreateCommand();
                _suburbCmd.CommandText = @"
                    SELECT name, place_type
                    FROM places
                    WHERE place_type IN ('suburb', 'neighbourhood', 'village', 'hamlet')
                      AND latitude BETWEEN @lat - 0.05 AND @lat + 0.05
                      AND longitude BETWEEN @lon - 0.05 AND @lon + 0.05
                    ORDER BY
                        (latitude - @lat) * (latitude - @lat) +
                        (longitude - @lon) * (longitude - @lon)
                    LIMIT 1";
                _suburbCmd.Parameters.Add("@lat", SqliteType.Real);
                _suburbCmd.Parameters.Add("@lon", SqliteType.Real);

                _cityCmd = _connection.CreateCommand();
                _cityCmd.CommandText = @"
                    SELECT name, place_type
                    FROM places
                    WHERE place_type IN ('city', 'town')
                      AND latitude BETWEEN @lat - 0.3 AND @lat + 0.3
                      AND longitude BETWEEN @lon - 0.3 AND @lon + 0.3
                    ORDER BY
                        (latitude - @lat) * (latitude - @lat) +
                        (longitude - @lon) * (longitude - @lon)
                    LIMIT 1";
                _cityCmd.Parameters.Add("@lat", SqliteType.Real);
                _cityCmd.Parameters.Add("@lon", SqliteType.Real);
            }
        }

        /// <summary>
        /// Load database bounds (called once during initialization)
        /// </summary>
        private DatabaseBounds LoadBounds()
        {
            using var cmd = _connection.CreateCommand();
            cmd.CommandText = @"
                SELECT
                    (SELECT CAST(value AS REAL) FROM metadata WHERE key = 'min_latitude'),
                    (SELECT CAST(value AS REAL) FROM metadata WHERE key = 'max_latitude'),
                    (SELECT CAST(value AS REAL) FROM metadata WHERE key = 'min_longitude'),
                    (SELECT CAST(value AS REAL) FROM metadata WHERE key = 'max_longitude'),
                    (SELECT CAST(value AS INTEGER) FROM metadata WHERE key = 'grid_size')";

            using var reader = cmd.ExecuteReader();
            if (reader.Read())
            {
                return new DatabaseBounds
                {
                    MinLatitude = reader.GetDouble(0),
                    MaxLatitude = reader.GetDouble(1),
                    MinLongitude = reader.GetDouble(2),
                    MaxLongitude = reader.GetDouble(3),
                    GridSize = reader.GetInt32(4)
                };
            }

            throw new Exception("Failed to load database bounds");
        }

        /// <summary>
        /// Calculate grid coordinates from GPS position
        /// </summary>
        private (int gridX, int gridY) CalculateGridCoords(double latitude, double longitude)
        {
            double normX = (longitude - _bounds.MinLongitude) /
                          (_bounds.MaxLongitude - _bounds.MinLongitude);
            double normY = (latitude - _bounds.MinLatitude) /
                          (_bounds.MaxLatitude - _bounds.MinLatitude);

            int gridX = (int)(normX * _bounds.GridSize);
            int gridY = (int)(normY * _bounds.GridSize);

            // Clamp to valid range
            gridX = Math.Max(0, Math.Min(_bounds.GridSize - 1, gridX));
            gridY = Math.Max(0, Math.Min(_bounds.GridSize - 1, gridY));

            return (gridX, gridY);
        }

        /// <summary>
        /// Get speed limit using grid-based lookup (FASTEST - use this)
        /// Returns speed limit in km/h, or -1 if not found
        /// Typical execution time: <1ms
        /// </summary>
        public int GetSpeedLimitGrid(double latitude, double longitude)
        {
            var (gridX, gridY) = CalculateGridCoords(latitude, longitude);

            _gridCommand.Parameters["@gridXMin"].Value = gridX - 1;
            _gridCommand.Parameters["@gridXMax"].Value = gridX + 1;
            _gridCommand.Parameters["@gridYMin"].Value = gridY - 1;
            _gridCommand.Parameters["@gridYMax"].Value = gridY + 1;
            _gridCommand.Parameters["@lat"].Value = latitude;
            _gridCommand.Parameters["@lon"].Value = longitude;

            var result = _gridCommand.ExecuteScalar();
            return result != null ? Convert.ToInt32(result) : -1;
        }

        /// <summary>
        /// Get speed limit using bounding box lookup (FALLBACK)
        /// Returns speed limit in km/h, or -1 if not found
        /// Typical execution time: <5ms
        /// </summary>
        public int GetSpeedLimitBBox(double latitude, double longitude)
        {
            _bboxCommand.Parameters["@lat"].Value = latitude;
            _bboxCommand.Parameters["@lon"].Value = longitude;

            var result = _bboxCommand.ExecuteScalar();
            return result != null ? Convert.ToInt32(result) : -1;
        }

        /// <summary>
        /// Get speed limit (tries grid first, falls back to bbox)
        /// Returns speed limit in km/h, or -1 if not found
        /// RECOMMENDED: Use this method for automatic fallback
        /// </summary>
        public int GetSpeedLimit(double latitude, double longitude)
        {
            int speedLimit = GetSpeedLimitGrid(latitude, longitude);

            // Fallback to bounding box if grid lookup fails
            if (speedLimit == -1)
            {
                speedLimit = GetSpeedLimitBBox(latitude, longitude);
            }

            return speedLimit;
        }

        /// <summary>
        /// Get detailed road information (use sparingly - more overhead)
        /// </summary>
        public RoadInfo? GetRoadInfo(double latitude, double longitude)
        {
            using var cmd = _connection.CreateCommand();
            cmd.CommandText = @"
                SELECT speed_limit_kmh, name, highway_type, is_inferred
                FROM road_segments
                WHERE center_lat BETWEEN @lat - 0.01 AND @lat + 0.01
                  AND center_lon BETWEEN @lon - 0.01 AND @lon + 0.01
                ORDER BY
                    (center_lat - @lat) * (center_lat - @lat) +
                    (center_lon - @lon) * (center_lon - @lon)
                LIMIT 1";

            cmd.Parameters.AddWithValue("@lat", latitude);
            cmd.Parameters.AddWithValue("@lon", longitude);

            using var reader = cmd.ExecuteReader();
            if (reader.Read())
            {
                return new RoadInfo
                {
                    SpeedLimitKmh = reader.GetInt32(0),
                    Name = reader.IsDBNull(1) ? null : reader.GetString(1),
                    HighwayType = reader.GetString(2),
                    IsInferred = reader.GetBoolean(3)
                };
            }

            return null;
        }

        /// <summary>
        /// Whether the database supports reverse geocoding (has places table)
        /// </summary>
        public bool HasLocationInfo => _hasPlacesTable;

        /// <summary>
        /// Get location info (street, suburb, city) for coordinates
        /// Returns null if places table is not available
        /// </summary>
        public LocationInfo? GetLocationInfo(double latitude, double longitude)
        {
            if (!_hasPlacesTable || _streetCmd == null || _suburbCmd == null || _cityCmd == null)
                return null;

            var info = new LocationInfo();

            _streetCmd.Parameters["@lat"].Value = latitude;
            _streetCmd.Parameters["@lon"].Value = longitude;
            using (var reader = _streetCmd.ExecuteReader())
            {
                if (reader.Read())
                    info.Street = reader.GetString(0);
            }

            _suburbCmd.Parameters["@lat"].Value = latitude;
            _suburbCmd.Parameters["@lon"].Value = longitude;
            using (var reader = _suburbCmd.ExecuteReader())
            {
                if (reader.Read())
                    info.Suburb = reader.GetString(0);
            }

            _cityCmd.Parameters["@lat"].Value = latitude;
            _cityCmd.Parameters["@lon"].Value = longitude;
            using (var reader = _cityCmd.ExecuteReader())
            {
                if (reader.Read())
                    info.City = reader.GetString(0);
            }

            return info;
        }

        private bool TableExists(string tableName)
        {
            using var cmd = _connection.CreateCommand();
            cmd.CommandText = "SELECT COUNT(*) FROM sqlite_master WHERE type='table' AND name=@name";
            cmd.Parameters.AddWithValue("@name", tableName);
            return Convert.ToInt64(cmd.ExecuteScalar()) > 0;
        }

        public void Dispose()
        {
            _streetCmd?.Dispose();
            _suburbCmd?.Dispose();
            _cityCmd?.Dispose();
            _gridCommand?.Dispose();
            _bboxCommand?.Dispose();
            _connection?.Dispose();
        }
    }

    public class LocationInfo
    {
        public string? Street { get; set; }
        public string? Suburb { get; set; }
        public string? City { get; set; }

        public override string ToString() =>
            $"{Street ?? "(unknown)"}, {Suburb ?? "(unknown)"}, {City ?? "(unknown)"}";
    }

    public class RoadInfo
    {
        public int SpeedLimitKmh { get; set; }
        public string? Name { get; set; }
        public string HighwayType { get; set; } = string.Empty;
        public bool IsInferred { get; set; }

        public override string ToString() =>
            $"{Name ?? "(unnamed)"} [{HighwayType}] {SpeedLimitKmh} km/h" +
            (IsInferred ? " (inferred)" : " (explicit)");
    }

    /* Example usage - COMMENTED OUT to avoid conflicts with main Program.cs

    class ProgramExample
    {
        static void Main()
        {
            using var lookup = new SpeedLimitLookup("Database/za_speedlimits.db");

            // Example 1: Simple lookup
            var speedLimit = lookup.GetSpeedLimit(-33.9249, 18.4241);
            Console.WriteLine($"Speed limit: {speedLimit} km/h");

            // Example 2: Detailed info
            var roadInfo = lookup.GetRoadInfo(-33.9249, 18.4241);
            if (roadInfo != null)
            {
                Console.WriteLine($"Road: {roadInfo}");
            }

            // Example 3: Batch lookups (reuses prepared statements)
            var locations = new[]
            {
                (-33.9249, 18.4241),  // Cape Town
                (-26.2041, 28.0473),  // Johannesburg
                (-33.9258, 18.4232)   // Residential
            };

            foreach (var (lat, lon) in locations)
            {
                var speed = lookup.GetSpeedLimit(lat, lon);
                Console.WriteLine($"({lat}, {lon}) -> {speed} km/h");
            }
        }
    }
    */
}
